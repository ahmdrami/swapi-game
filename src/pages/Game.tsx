import { useQuery } from '@apollo/client'
import { Fragment } from 'react'
import { useHistory, useParams } from 'react-router-dom'
import { Box, Button, Flex, Heading, Spinner } from 'theme-ui'

import { Card, Layout } from '../components'
import { QUERY_PERSON_TOTAL, QUERY_STARSHIPS_TOTAL } from '../gql/queries'
import { useNewGame } from '../hooks/useNewGame'

type Params = {
  profile: 'person' | 'starship'
}

export const Game = () => {
  const { profile } = useParams<Params>()
  const history = useHistory()

  const allEntries = useQuery(
    profile === 'person' ? QUERY_PERSON_TOTAL : QUERY_STARSHIPS_TOTAL
  )
  const totalCount = allEntries.data
    ? allEntries.data[Object.keys(allEntries.data)[0]].totalCount
    : 0

  const { data, loading, initNewGame, winner } = useNewGame(
    profile,
    2,
    totalCount
  )
  return (
    <Layout>
      {loading ? (
        <Spinner />
      ) : (
        <Fragment>
          <Flex
            sx={{
              justifyContent: 'center',
              alignItems: 'center',
            }}
          >
            {data.map((card) => (
              <Card {...card} key={card.id} />
            ))}
          </Flex>
          <Flex
            sx={{
              justifyContent: 'center',
              alignItems: 'center',
              flexDirection: 'column',
              mt: 5,
            }}
          >
            <Heading
              as="h2"
              variant="styled.h2"
              sx={{
                textTransform: 'capitalize',
                p: 3,
                borderRadius: '5px',
              }}
            >
              {!winner ? 'Draw' : `${winner} wins`}
            </Heading>
            <Box
              sx={{
                mt: 4,
                '> button:not(:last-child)': {
                  mr: 2,
                },
              }}
            >
              <Button variant="primary" onClick={() => initNewGame(true)}>
                Play again
              </Button>
              <Button variant="secondary" onClick={() => history.push('/')}>
                New Game
              </Button>
            </Box>
          </Flex>
        </Fragment>
      )}
    </Layout>
  )
}
