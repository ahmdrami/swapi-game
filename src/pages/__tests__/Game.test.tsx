import { render } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
import { Game } from '../Game'
import { MockedProvider } from '@apollo/client/testing'

describe('Game page component', () => {
  test('should render the fragment', () => {
    const { asFragment } = render(
      <MemoryRouter>
        <MockedProvider>
          <Game />
        </MockedProvider>
      </MemoryRouter>
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
