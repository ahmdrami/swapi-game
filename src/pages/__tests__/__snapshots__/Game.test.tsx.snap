// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game page component should render the fragment 1`] = `
<DocumentFragment>
  <div
    class="css-16zix2c"
  >
    <svg
      class="css-1pf5xd"
      fill="none"
      height="48"
      role="img"
      stroke="currentcolor"
      stroke-width="4"
      viewBox="0 0 32 32"
      width="48"
    >
      <title>
        Loading...
      </title>
      <circle
        cx="16"
        cy="16"
        opacity="0.125"
        r="12"
      />
      <circle
        class="css-wpcq6n"
        cx="16"
        cy="16"
        r="12"
        stroke-dasharray="75.39822368615503"
        stroke-dashoffset="56.548667764616276"
      />
    </svg>
  </div>
</DocumentFragment>
`;
